File: packages/react-native-client/ios/RNFishjamClientModule.swift
import ExpoModulesCore

struct RNSimulcastConfig: Record {
    @Field
    var enabled: Bool = false
}

struct CameraConfig: Record {
    @Field
    var quality: String = "VGA169"

    @Field
    var flipDimensions: Bool = false

    @Field
    var videoTrackMetadata: [String: Any] = [:]

    @Field
    var simulcastConfig: RNSimulcastConfig = RNSimulcastConfig()

    @Field
    var cameraEnabled: Bool = true

    @Field
    var cameraId: String? = nil
}

struct MicrophoneConfig: Record {
    @Field
    var audioTrackMetadata: [String: Any] = [:]

    @Field
    var microphoneEnabled: Bool = true
}

struct ScreenShareOptions: Record {
    @Field
    var quality: String = "HD15"

    @Field
    var screenShareMetadata: [String: Any] = [:]

    @Field
    var simulcastConfig: RNSimulcastConfig = RNSimulcastConfig()
}

struct ReconnectConfig: Record {
    @Field
    var maxAttempts: Int = 5

    @Field
    var initialDelayMs: Int = 1000

    @Field
    var delayMs: Int = 1000
}

struct ConnectConfig: Record {
    @Field
    var reconnectConfig: ReconnectConfig = ReconnectConfig()
}

typealias RNTrackBandwidthLimit = Either<Int, [String: Int]>

public class RNFishjamClientModule: Module {
  public func definition() -> ModuleDefinition {
    let rnFishjamClient: RNFishjamClient = {
      let client = RNFishjamClient(sendEvent)
      client.create()
      return client
    }()

    Name("RNFishjamClient")

    Events(EmitableEvent.allEvents)
    
    OnCreate {
          let permissionsManager = self.appContext?.permissions
          EXPermissionsMethodsDelegate.register(
            [
              CameraOnlyPermissionRequester(),
              CameraMicrophonePermissionRequester()
            ],
            withPermissionsManager: permissionsManager
          )
        }

    Property("peerStatus") {
      return rnFishjamClient.peerStatus.rawValue
    }

    Property("reconnectionStatus") {
      return rnFishjamClient.reconnectionStatus.rawValue
    }

    Property("isMicrophoneOn") {
      return rnFishjamClient.isMicrophoneOn
    }

    Property("isCameraOn") {
      return rnFishjamClient.isCameraOn
    }

    Property("cameras") {
      return rnFishjamClient.getCaptureDevices()
    }

    Property("currentCamera") {
      return rnFishjamClient.currentCamera
    }

    Property("isScreenShareOn") {
      return rnFishjamClient.isScreenShareOn
    }

    Property("isAppScreenShareOn") {
      return rnFishjamClient.isAppScreenShareOn
    }

    Property("isCameraInitialized") {
      return rnFishjamClient.isCameraInitialized
    }

    Function("getPeers") {
      return rnFishjamClient.getPeers()
    }

    AsyncFunction("joinRoom") {
      (
        url: String, peerToken: String, peerMetadata: [String: Any], config: ConnectConfig,
        promise: Promise
      ) in
      rnFishjamClient.joinRoom(
        url: url, peerToken: peerToken, peerMetadata: peerMetadata, config: config,
        promise: promise)
    }

    AsyncFunction("leaveRoom") {
      rnFishjamClient.leaveRoom()
    }

    AsyncFunction("startCamera") { (config: CameraConfig) in
      try await rnFishjamClient.startCamera(config: config)
    }

    AsyncFunction("toggleMicrophone") {
      try await rnFishjamClient.toggleMicrophone()
    }
    
    AsyncFunction("startMicrophone") {
      try await rnFishjamClient.startMicrophone()
    }

    AsyncFunction("stopMicrophone") {
      try rnFishjamClient.stopMicrophone()
    }

    AsyncFunction("toggleCamera") {
      try rnFishjamClient.toggleCamera()
    }

    AsyncFunction("flipCamera") {
      try rnFishjamClient.flipCamera()
    }

    AsyncFunction("switchCamera") { (cameraId: String) in
      try rnFishjamClient.switchCamera(cameraId: cameraId)
    }

    AsyncFunction("toggleScreenShare") { (screenShareOptions: ScreenShareOptions) in
      try rnFishjamClient.toggleScreenShare(screenShareOptions: screenShareOptions)
    }

    AsyncFunction("toggleAppScreenShare") { (screenShareOptions: ScreenShareOptions) in
      try rnFishjamClient.toggleAppScreenShare(screenShareOptions: screenShareOptions)
    }

    AsyncFunction("updatePeerMetadata") { (metadata: [String: Any]) in
      try rnFishjamClient.updatePeerMetadata(metadata: metadata)
    }

    AsyncFunction("updateVideoTrackMetadata") { (metadata: [String: Any]) in
      try rnFishjamClient.updateLocalVideoTrackMetadata(metadata: metadata)
    }

    AsyncFunction("updateScreenShareTrackMetadata") { (metadata: [String: Any]) in
      try rnFishjamClient.updateLocalScreenShareTrackMetadata(metadata: metadata)
    }

    AsyncFunction("toggleScreenShareTrackEncoding") { (encoding: String) in
      try rnFishjamClient.toggleScreenShareTrackEncoding(encoding: encoding)
    }

    AsyncFunction("setScreenShareTrackBandwidth") { (bandwidth: Int) in
      try rnFishjamClient.setScreenShareTrackBandwidth(bandwidth: bandwidth)
    }

    AsyncFunction("setScreenShareTrackEncodingBandwidth") { (encoding: String, bandwidth: Int) in
      try rnFishjamClient.setScreenShareTrackEncodingBandwidth(
        encoding: encoding, bandwidth: bandwidth)
    }

    AsyncFunction("setTargetTrackEncoding") { (trackId: String, encoding: String) in
      try rnFishjamClient.setTargetTrackEncoding(trackId: trackId, encoding: encoding)
    }

    AsyncFunction("toggleVideoTrackEncoding") { (encoding: String) in
      try rnFishjamClient.toggleVideoTrackEncoding(encoding: encoding)
    }

    AsyncFunction("setVideoTrackEncodingBandwidth") { (encoding: String, bandwidth: Int) in
      try rnFishjamClient.setVideoTrackEncodingBandwidth(
        encoding: encoding, bandwidth: bandwidth)
    }

    AsyncFunction("setVideoTrackBandwidth") { (bandwidth: Int) in
      try rnFishjamClient.setVideoTrackBandwidth(bandwidth: bandwidth)
    }

    AsyncFunction("changeWebRTCLoggingSeverity") { (severity: String) in
      try rnFishjamClient.changeWebRTCLoggingSeverity(severity: severity)
    }

    AsyncFunction("getStatistics") {
      try await rnFishjamClient.getStatistics()
    }

    AsyncFunction("selectAudioSessionMode") { (sessionMode: String) in
      try rnFishjamClient.selectAudioSessionMode(sessionMode: sessionMode)
    }

    AsyncFunction("showAudioRoutePicker") {
      rnFishjamClient.showAudioRoutePicker()
    }

    AsyncFunction("startAudioSwitcher") {
      rnFishjamClient.startAudioSwitcher()
    }

    AsyncFunction("getCameraPermissionsAsync") { (promise: Promise) in
      EXPermissionsMethodsDelegate.getPermissionWithPermissionsManager(
        self.appContext?.permissions,
        withRequester: CameraOnlyPermissionRequester.self,
        resolve: promise.resolver,
        reject: promise.legacyRejecter
      )
    }

    AsyncFunction("requestCameraPermissionsAsync") { (promise: Promise) in
      EXPermissionsMethodsDelegate.askForPermission(
        withPermissionsManager: self.appContext?.permissions,
        withRequester: CameraOnlyPermissionRequester.self,
        resolve: promise.resolver,
        reject: promise.legacyRejecter
      )
    }

    AsyncFunction("getMicrophonePermissionsAsync") { (promise: Promise) in
      EXPermissionsMethodsDelegate.getPermissionWithPermissionsManager(
        self.appContext?.permissions,
        withRequester: CameraMicrophonePermissionRequester.self,
        resolve: promise.resolver,
        reject: promise.legacyRejecter
      )
    }

    AsyncFunction("requestMicrophonePermissionsAsync") { (promise: Promise) in
      EXPermissionsMethodsDelegate.askForPermission(
        withPermissionsManager: self.appContext?.permissions,
        withRequester: CameraMicrophonePermissionRequester.self,
        resolve: promise.resolver,
        reject: promise.legacyRejecter
      )
    }

    }
  }

File: packages/react-native-client/android/src/main/java/io/fishjam/reactnative/RNFishjamClientModule.kt
package io.fishjam.reactnative

import expo.modules.kotlin.Promise
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

class SimulcastConfig : Record {
  @Field
  val enabled: Boolean = false
}

class CameraConfig : Record {
  @Field
  val quality: String = "VGA169"

  @Field
  val flipDimensions: Boolean = false

  @Field
  val videoTrackMetadata: Map<String, Any> = emptyMap()

  @Field
  val simulcastConfig: SimulcastConfig = SimulcastConfig()

  @Field
  val cameraEnabled: Boolean = true

  @Field
  val cameraId: String? = null
}

class ScreenShareOptions : Record {
  @Field
  val quality: String = "HD15"

  @Field
  val screenShareMetadata: Map<String, Any> = emptyMap()

  @Field
  val simulcastConfig: SimulcastConfig = SimulcastConfig()
}

class ReconnectConfig : Record {
  @Field
  val maxAttempts: Int = 5

  @Field
  val initialDelayMs: Long = 1000

  @Field
  val delayMs: Long = 1000
}

class ConnectConfig : Record {
  @Field
  val reconnectConfig: ReconnectConfig = ReconnectConfig()
}

class ForegroundServiceConfig : Record {
  @Field
  val channelId: String? = null

  @Field
  val channelName: String? = null

  @Field
  val notificationContent: String? = null

  @Field
  val notificationTitle: String? = null

  @Field
  val enableCamera: Boolean? = null

  @Field
  val enableMicrophone: Boolean? = null
}

class RNFishjamClientModule : Module() {
  override fun definition() =
    ModuleDefinition {
      val mutex = Mutex()

      val rnFishjamClient =
        RNFishjamClient { name: String, data: Map<String, Any?> ->
          sendEvent(name, data)
        }

      Name("RNFishjamClient")

      Events(EmitableEvent.allEvents)

      OnCreate {
        rnFishjamClient.onModuleCreate(appContext)
      }

      OnDestroy {
        rnFishjamClient.onModuleDestroy()
      }

      OnActivityDestroys {
        rnFishjamClient.leaveRoom()
      }

      OnActivityResult { _, result ->
        rnFishjamClient.onActivityResult(result.requestCode, result.resultCode, result.data)
      }

      Property("isCameraOn") {
        return@Property rnFishjamClient.isCameraOn
      }

      Property("isMicrophoneOn") {
        return@Property rnFishjamClient.isMicrophoneOn
      }

      Property("cameras") {
        return@Property rnFishjamClient.getCaptureDevices()
      }

      Property("currentCamera") {
        return@Property rnFishjamClient.getCurrentCaptureDevice()
      }

      Property("isScreenShareOn") {
        return@Property rnFishjamClient.isScreenShareOn
      }

      Property("peerStatus") {
        return@Property rnFishjamClient.peerStatus.status
      }

      Property("reconnectionStatus") {
        return@Property rnFishjamClient.reconnectionStatus.status
      }

      Property("isCameraInitialized") {
        return@Property rnFishjamClient.isCameraInitialized
      }

      Function("getPeers") {
        return@Function rnFishjamClient.getPeers()
      }

      AsyncFunction(
        "joinRoom"
      ) { url: String, peerToken: String, peerMetadata: Map<String, Any>, config: ConnectConfig, promise: Promise ->
        CoroutineScope(Dispatchers.Main).launch {
          rnFishjamClient.joinRoom(url, peerToken, peerMetadata, config, promise)
        }
      }

      AsyncFunction("leaveRoom") Coroutine { ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.leaveRoom()
        }
      }

      AsyncFunction("startCamera") Coroutine { config: CameraConfig ->
        return@Coroutine withContext(Dispatchers.Main) {
          return@withContext rnFishjamClient.startCamera(config)
        }
      }

      AsyncFunction("toggleMicrophone") Coroutine { ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.toggleMicrophone()
        }
      }

      AsyncFunction("toggleCamera") Coroutine { ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.toggleCamera()
        }
      }

      AsyncFunction("flipCamera") Coroutine { ->
        mutex.withLock {
          withContext(Dispatchers.Main) {
            rnFishjamClient.flipCamera()
          }
        }
      }

      AsyncFunction("switchCamera") Coroutine { cameraId: String ->
        mutex.withLock {
          withContext(Dispatchers.Main) {
            rnFishjamClient.switchCamera(cameraId)
          }
        }
      }

      AsyncFunction("handleScreenSharePermission") { promise: Promise ->
        CoroutineScope(Dispatchers.Main).launch {
          rnFishjamClient.handleScreenSharePermission(promise)
        }
      }

      AsyncFunction("toggleScreenShare") Coroutine { screenShareOptions: ScreenShareOptions ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.toggleScreenShare(screenShareOptions)
        }
      }

      AsyncFunction("updatePeerMetadata") Coroutine { metadata: Map<String, Any> ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.updatePeerMetadata(metadata)
        }
      }

      AsyncFunction("updateVideoTrackMetadata") Coroutine { metadata: Map<String, Any> ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.updateLocalVideoTrackMetadata(metadata)
        }
      }

      AsyncFunction("updateScreenShareTrackMetadata") Coroutine { metadata: Map<String, Any> ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.updateLocalScreenShareTrackMetadata(metadata)
        }
      }

      AsyncFunction("setOutputAudioDevice") { audioDevice: String ->
        rnFishjamClient.setOutputAudioDevice(audioDevice)
      }

      AsyncFunction("startAudioSwitcher") {
        rnFishjamClient.startAudioSwitcher()
      }

      AsyncFunction("stopAudioSwitcher") {
        rnFishjamClient.stopAudioSwitcher()
      }

      AsyncFunction("toggleScreenShareTrackEncoding") Coroutine { encoding: String ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.toggleScreenShareTrackEncoding(encoding)
        }
      }

      AsyncFunction("setScreenShareTrackBandwidth") Coroutine { bandwidth: Int ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.setScreenShareTrackBandwidth(bandwidth)
        }
      }

      AsyncFunction("setScreenShareTrackEncodingBandwidth") Coroutine { encoding: String, bandwidth: Int ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.setScreenShareTrackEncodingBandwidth(encoding, bandwidth)
        }
      }

      AsyncFunction("setTargetTrackEncoding") Coroutine { trackId: String, encoding: String ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.setTargetTrackEncoding(trackId, encoding)
        }
      }

      AsyncFunction("toggleVideoTrackEncoding") Coroutine { encoding: String ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.toggleVideoTrackEncoding(encoding)
        }
      }

      AsyncFunction("setVideoTrackEncodingBandwidth") Coroutine { encoding: String, bandwidth: Int ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.setVideoTrackEncodingBandwidth(encoding, bandwidth)
        }
      }

      AsyncFunction("setVideoTrackBandwidth") Coroutine { bandwidth: Int ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.setVideoTrackBandwidth(bandwidth)
        }
      }

      AsyncFunction("changeWebRTCLoggingSeverity") Coroutine { severity: String ->
        withContext(Dispatchers.Main) {
          rnFishjamClient.changeWebRTCLoggingSeverity(severity)
        }
      }

      AsyncFunction("getStatistics") Coroutine { ->
        return@Coroutine rnFishjamClient.getStatistics()
      }

      AsyncFunction("startForegroundService") Coroutine { config: ForegroundServiceConfig ->
        rnFishjamClient.startForegroundService(config)
      }

      Function("stopForegroundService") {
        rnFishjamClient.stopForegroundService()
      }
    }
}
